File: 0044-wildcard-matching/0044-wildcard-matching.cpp
class Solution {
public:
    bool match(int i, int j, string &s, string &p, vector<vector<int>> &dp) {
        // both string and pattern finished
        if (i < 0 && j < 0) return true;

        // string finished but pattern left
        if (i < 0 && j >= 0) {
            for (int k = 0; k <= j; k++) {
                if (p[k] != '*') return false;
            }
            return true;
        }

        // pattern finished but string left
        if (i >= 0 && j < 0) return false;

        // check dp
        if (dp[i][j] != -1) return dp[i][j];

        bool ans = false;

        // character matches or '?'
        if (s[i] == p[j] || p[j] == '?') {
            ans = match(i - 1, j - 1, s, p, dp);
        }
        else if (p[j] == '*') {
            ans = match(i - 1, j, s, p, dp)   // '*' matches one char
               || match(i, j - 1, s, p, dp); // '*' matches empty
        }

        return dp[i][j] = ans;
    }

    bool isMatch(string s, string p) {
        int n = s.size();
        int m = p.size();
        vector<vector<int>> dp(n, vector<int>(m, -1));
        return match(n - 1, m - 1, s, p, dp);
    }
};

File: 0044-wildcard-matching/0044-wildcard-matching.java
class Solution {
    public boolean isMatch(String s, String p) {
        char[] str=s.toCharArray();
        char[] pattern=p.toCharArray();
        int writeIndex=0;
        boolean isFirst=true;
        for(int i=0;i<pattern.length;i++){
            if(pattern[i]=='*'){
                if(isFirst){
                    pattern[writeIndex++]= pattern[i];
                    isFirst=false;
                }
            }
            else{
                pattern[writeIndex++]=pattern[i];
                isFirst=true;
            }
        }
        boolean[][] dp=new boolean[str.length+1][writeIndex+1];
        dp[0][0]=true;
        for(int j=1;j<=writeIndex;j++){
            if(pattern[j-1]=='*'){
                dp[0][j]=dp[0][j-1];
            }
        }
        for(int i=1;i<=str.length;i++){
            for(int j=1;j<=writeIndex;j++){
                if(pattern[j-1]=='?' || str[i-1]==pattern[j-1]){
                    dp[i][j]=dp[i-1][j-1];
                }
                else if(pattern[j-1]=='*'){
                    dp[i][j]=dp[i-1][j]|| dp[i][j-1];
                }
            }
        }
        return dp[str.length][writeIndex];
    }
}